File: 2023-2024/sourcecode/mod3/RetrofitBasicsJava/app/src/main/java/com/example/retrofitbasicsjava/data/api/RetrofitInstance.java
package com.example.retrofitbasicsjava.data.api;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitInstance {
    private static final String BASE_URL = "https://jsonplaceholder.typicode.com/";
    private static volatile PlaceholderApi apiInstance;

    public static PlaceholderApi getApi() {
        if (apiInstance == null) {
            synchronized (RetrofitInstance.class) {
                if (apiInstance == null) {
                    apiInstance = new Retrofit.Builder()
                            .baseUrl(BASE_URL)
                            .addConverterFactory(GsonConverterFactory.create())
                            .build()
                            .create(PlaceholderApi.class);
                }
            }
        }
        return apiInstance;
    }
}

File: 2023-2024/sourcecode/mod4/lab12/ResourcePatternBasicsKotlin/app/src/main/java/com/example/resourcepatternbasicskotlin/data/api/RetrofitInstance.kt
package com.example.resourcepatternbasicskotlin.data.api

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitInstance {
    val api: PlaceholderApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://jsonplaceholder.typicode.com/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(PlaceholderApi::class.java)
    }
}